src/routers/Gateway.sol

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.18;
+pragma abicoder v2;
+
+import {TypedMemView} from "memview-sol/TypedMemView.sol";
+import {ConnectorMessages} from "../Messages.sol";
+
+interface ConnectorLike {
+    function addPool(uint64 poolId, uint128 currency, uint8 decimals) external;
+    function addTranche(
+        uint64 poolId,
+        bytes16 trancheId,
+        string memory tokenName,
+        string memory tokenSymbol,
+        uint128 price
+    ) external;
+    function updateMember(uint64 poolId, bytes16 trancheId, address user, uint64 validUntil) external;
+    function updateTokenPrice(uint64 poolId, bytes16 trancheId, uint128 price) external;
+    function handleTransferTrancheTokens(
+        uint64 poolId,
+        bytes16 trancheId,
+        uint256 destinationChainId,
+        address destinationAddress,
+        uint128 amount
+    ) external;
+}
+
+interface RouterLike {
+    function send(bytes memory message) external;
+}
+
+contract ConnectorGateway {
+    using TypedMemView for bytes;
+    // why bytes29? - https://github.com/summa-tx/memview-sol#why-bytes29
+    using TypedMemView for bytes29;
+    using ConnectorMessages for bytes29;
+
+    mapping(address => uint256) public wards;
+
+    ConnectorLike public immutable connector;
+    // TODO: support multiple incoming routers (just a single outgoing router) to simplify router migrations
```

Comment:
@offerijns / @AStox I am curious as to how we would go about implementing this

Reply:
I think it can be as simple as having a `mapping (address => bool) routers` and then adapting `onlyRouter` to check that the `msg.sender` is in this map.